calico/openstack/mech_calico.py

```diff
@@ -196,14 +206,148 @@ def _init_state(self):
+        """
+        try:
+            self._client = etcd.Client(host=cfg.CONF.calico.etcd_host,
+                                       port=cfg.CONF.calico.etcd_port)
+            while True:
+                # Get and handle an update.
+                response = self._client.read(STATUS_DIR,
+                                             wait=True,
+                                             waitIndex = self._next_etcd_index,
+                                             recursive=True)
+
+                # Before status is passed to Neutron db, check whether we
+                # are still master (if not, we loop until we become one)
+                # and whether epoch is as expected (if not, thread exits).
+                if self._epoch != expected_epoch or not self.transport.is_master:
+                    break
+
+                self._handle_status_update(response)
+                # Since we're polling on a subtree, we can't just
+                # increment the index, we have to look at the
+                # modifiedIndex.
+                self._next_etcd_index = max(self._next_etcd_index,
+                                           int(response.modifiedIndex)) + 1
+        except (ReadTimeoutError, SocketTimeout) as e:
+            # This is expected when we're doing a poll and nothing
+            # happened. socket timeout doesn't seem to be caught by
+            # urllib3 1.7.1. Simply reconnect.
+            LOG.debug("Read from etcd timed out (%r), retrying.", e)
+        except (etcd.EtcdClusterIdChanged, etcd.EtcdEventIndexCleared) as e:
+            # Complete resync is needed.
+            LOG.warning("Out of sync with etcd (%r).  Reconnecting "
+                     "for full sync.", e)
+            try:
+                context = ctx.get_admin_context()
+                self.complete_resync(context)
+            except Exception:
+                LOG.exception("Error in full resync.")
+        except Exception as e:
+            LOG.info("Handling status update failed. ({})".format(e))
+
+    def _handle_status_update(self, response):
+        """
+        Handle updates of felix instances written to etcd. Pass status
+        updates to neutron database.
+
+        :param response: update written to etcd by one of felix instances
+        """
+        # Only get info from uptime keys - since those have ttl, we know
+        # if key was updated or created.
+        key = response.key
+        if not is_uptime_key(key):
+            return
+        # Ignore deleted and expired keys - their deletion is noticed by
+        # other timeout algorithm before reaching the horizon UI.
+        if response.action == "delete" or response.action == "expire":
+            return
+
+
+        hostname = _hostname_from_status_key(key)
+        agent_state = {'agent_type': AGENT_TYPE_FELIX,
+                       'binary': '',
+                       'host': hostname,
+                       'topic': constants.L2_AGENT_TOPIC}
```

Comment:
Hmm, seems fishy that we use the L2_AGENT_TOPIC, when Calico is an L3 solution.  What is this used for.  

CC @neiljerram